morph_tool/morphdb.py

```diff
@@ -0,0 +1,257 @@
+    and methods to write neurondb to various format (xml, dat, csv)
+    '''
+
+    def __init__(self,
+                 neurondb: Path = None,
+                 morph_info_filter: Callable[[MorphInfo], bool] = None):
+        '''Builds a MorphologyDB from a neurondb.xml file
+
+        Args:
+            neurondb: path to a neurondb.xml file
+            morph_info_filter: a filter function to be applied to each MorphInfo element
+        '''
+        self.lineage = {}
+
+        if neurondb is not None:
+            with open(neurondb) as fd:
+                neurondb = xmltodict.parse(fd.read())
+
+            morphologies = neurondb['neurondb']['listing']['morphology']
+
+            # Case where there is a single <morphology></morphology> tag in the XML
+            if not isinstance(morphologies, list):
+                morphologies = [morphologies]
+            morphologies = filter(None, morphologies)
+            morphologies = map(MorphInfo, morphologies)
+            if morph_info_filter is not None:
+                morphologies = filter(morph_info_filter, morphologies)
+        else:
+            morphologies = []
+
+        self.morphologies = list(morphologies)
+        self._known_morphologies = set(map(hash, self.morphologies))
+
+    def __iter__(self):
+        return iter(self.morphologies)
+
+    def remove_morphs(self, removed_morphs: Iterable[MorphInfo]) -> None:
+        '''Removes morphologies from the database.'''
+        removed_morphs = set(removed_morphs)
+        self.morphologies = [morph for morph in self.morphologies
+                             if morph.name not in removed_morphs]
+
+        self._known_morphologies = set(map(hash, self.morphologies))
+
+        for name in removed_morphs:
+            self.lineage.pop(name, None)
+
+    def add_morph(self, morph_info: MorphInfo) -> bool:
+        '''Add a morphology to the database.
+
+        If the name, mtype, layer combination already exists, then nothing is added
+
+        Return:
+            True if morphology added, False otherwise
+        '''
+        key = hash(morph_info)
+        if key in self._known_morphologies:
+            return False
+        self._known_morphologies.add(key)
+        self.morphologies.append(morph_info)
+
+        lineage = {'dendrite': morph_info.dendrite_donor, 'axon': morph_info.axon_donor}
+        lineage = {k: v for k, v in lineage.items() if v is not None}
+        if lineage:
+            self.lineage[morph_info.name] = lineage
+        return True
+
+    @property
+    def df(self):
+        '''Returns a pandas.DataFrame view of the data with the following columns:
+
+           'name': the morpho name (without extension)
+           'mtype': the mtype (without msubtype),
+           'msubtype': the msubtype (the part of the fullmtype after the ":")
+           'fullmtype': the full mtype (mtype:msubtype)
+           'layer': the layer (as a string)
```

Comment:
I'm tempted to change the name of this column to `region`. That would makes for sense for neurondb of hippocampus and thalamus.

Reply:
`region` usually applies to things like `MOs`, `FRM`, etc, so leaving it as layer is not optimal, but makes sense.  I've used `subregion` sometimes - but I don't think that's always clear.  I would poll the scientists, to see what they would like.

Reply:
Elisabetta says in the case in the thalamus what appears under the `layer` tag are regions. I'm still waiting for Davide's answer.

> region usually applies to things like MOs, FRM

What are MOs, FRM and where can I see those terms ? Actual region from SSCX ?

Reply:
I would be more in favor or leaving the standard `layer` tag, as otherwise, most users will have to remember to change it  to `layer`, and people using `layer` as `region` are more aware of it that people using `layer` as `layer` (I think!)

Reply:
The AIBS hierarchy looks like:
```
  "acronym": "FRP",                                                  
  "id": 184,                                                         
  "name": "Frontal pole, cerebral cortex",
  "children": [{
    "acronym": "FRP1",                                               
    "id": 68,                                                        
    "name": "Frontal pole, layer 1"                                  
   },                                                                
   {                                                                 
    "acronym": "FRP2",                                               
    "id": 20667,                                                     
    "name": "Frontal pole, layer 2"                                  
   },                                                                
  ....]
```
And, so far, we've considered 'FRP' to be a region, and its layers to be [1,2, 3..]: ie: leaf nodes are `layers` (`subregion`?), and regions are one level up.

The thalamus looks like:
```
  {                                                                 
   "acronym": "VP",                                                 
   "id": 709,                                                       
   "name": "Ventral posterior complex of the thalamus"              
   "children": [                                                    
    {                                                               
     "acronym": "VPL",                                              
     "id": 718,                                                     
     "name": "Ventral posterolateral nucleus of the thalamus"       
    },                                                              
    {                                                               
     "acronym": "VPLpc",                                            
     "id": 725,                                                     
     "name": "Ventral posterolateral nucleus of the thalamus, parvicellular part"
    },                                                              
    {                                                               
     "acronym": "VPM",                                              
     "id": 733,                                                     
     "name": "Ventral posteromedial nucleus of the thalamus"        
    },                                                              
    {                                                               
     "acronym": "VPMpc",                                            
     "id": 741,                                                     
     "name": "Ventral posteromedial nucleus of the thalamus, parvicellular part"
    }                                                               
   ],                                                               
  }, 
```
So, this *could* be considered to have region='VP', w/ `layers` / `subregion`s `VPL/VPM/...`

However, from what I remember from Elisabetta, she usually has `region` = VPL, and then would have no `subregion`.

This is to give a sense of the picture - maybe DKE (@annakristinkaufmann or @MFSY) can help w/ the terminology?  It would be good to use the standardization / create one.

Reply:
I guess (you probably know more about this Mike) that what is found in the current `layer` tag of the XML is the maximum amount of knowledge we have about the location of the morphology in the current state of things. That is, it could be anything between just a layer number and an actual region-subregion combo. In short I don't think the current neurondb.xml data we currently have are sufficient to correctly define what is expected in the `layer` tag.

So:
- we keep using `layer`. As Alexis said, non-sscx people are already used to this (for them) non-precise terminology. Maybe it's not the time to start changing the terminology. Moreover, this API has an intermediate lifetime: once Nexus is widespread neurondb.xml and the API will probably become obselete.
- we use `region` in the broader sense of things. After all, `layer 1`, `VP` or `VPL` can be considered 'brain regions'.